### PR TITLE
Begin using `curies.Converter` in more places

### PR DESCRIPTION
**Pull request by cthoyt**
Part of #363 

This PR does the following:

1. Adds a minimum version of `curies` that has the strict compress and expand functions
2. Rewrites the SPARQL utils and RDF utils to use `curies` functionality
3. Updates custom `curie_from_uri` to use `curies` (will make a follow-up PR that replaces this completely)

### CODE REVIEW

File: src/sssom/cli.py
Comment (matentzn) [inline src/sssom/cli.py:359]: Nice syntax!

File: src/sssom/sparql_util.py
Comment (matentzn) [inline src/sssom/sparql_util.py:67]: Cool method, textwrap.dedent!

File: src/sssom/sparql_util.py
Comment (matentzn) [inline src/sssom/sparql_util.py:99]: I am so glad this is going away.

File: src/sssom/util.py
Comment (matentzn) [inline src/sssom/util.py:1110]: Add deprecation annotation?

File: src/sssom/util.py
Comment (matentzn) [inline src/sssom/util.py:1131]: What happens if `prefix_map` _is_ an instance of `converter`? I mean, this case should basically never happen, but shouldn't there be an

```
else: converter = prefix_map 
```

There?

File: src/sssom/util.py
Comment (matentzn) [inline src/sssom/util.py:1605]: Is there a reason the ValueError is not a bit more verbose?

**Review by matentzn — DISMISSED** [review]
Looks awesome!! THANK YOU!

@hrshdhgd I will leave it to you to deal with the PR - I left 1/2 comments here and there to check, but nothing critical.

File: src/sssom/util.py
Comment (cthoyt) [reply src/sssom/util.py:1131]: oop this is a typo, thank you for the catch.

File: src/sssom/util.py
Comment (cthoyt) [reply src/sssom/util.py]: Fixed in 7e19639

File: src/sssom/util.py
Comment (cthoyt) [reply src/sssom/util.py]: Fixed in ec3452f

File: src/sssom/util.py
Comment (cthoyt) [reply src/sssom/util.py:1113]: I wasn't aware of this, thanks for the suggestion. Done in 118017d.

**Comment by cthoyt** [comment]
@matentzn @hrshdhgd it appears the issues in the build now are due to poetry. I will again give a gentle nudge to totally drop poetry, it is way more of an issue than it helps

**Comment by hrshdhgd** [comment]
There are 4 errors as of now. First one is a CLI error which will be fixed automatically once the following are fixed.

Three pytest errors occurring as of now:
```
FAILED tests/test_parsers.py::TestParse::test_parse_alignment_minidom - curies.api.DuplicateURIPrefixes: Duplicate URI prefixes:
FAILED tests/test_parsers.py::TestParse::test_parse_obographs - curies.api.DuplicateURIPrefixes: Duplicate URI prefixes:
FAILED tests/test_parsers.py::TestParse::test_parse_sssom_rdf - ValueError: CURIE appeared where there should be a URI, and could not be standardized: oio:hasBroadSynonym
```

From what I can tell the first two errors are thrown by `curies/api.py`.

The third error is because the prefix `oio:` is not in the `prefix_map`.

**Comment by hrshdhgd** [comment]
>it appears the issues in the build now are due to poetry. I will again give a gentle nudge to totally drop poetry, it is way more of an issue than it helps

I agree but it seems to me that only GitHub environments seem to be problematic. Locally, it seems to work fine. Plus Chris insists all our projects following a standard architecture, hence my hands are sort of tied 🙂 .

**Comment by cthoyt** [comment]
@hrshdhgd thanks for following up on all of this. can you take care of looking into these tests that have inconsistent data? Great to see that we are implicitly now checking data integrity using `curies`!

**Comment by hrshdhgd** [comment]
I've fixed the `poetry` issue @cthoyt . You may be in a position to fix these errors quicker than I and I have other higher priorities to tend to atm. I'd really appreciate if you could complete your PR.

File: src/sssom/context.py
Comment (matentzn) [inline src/sssom/context.py:90]: I dont know who wrote this but this seems misplaced. Why is this here?

File: src/sssom/sparql_util.py
Comment (matentzn) [inline src/sssom/sparql_util.py:38]: Maybe a more expressive error message here

Comment (cthoyt) [reply src/sssom/sparql_util.py]: Done in 1c25c94, but I inherited this not having a useful error message from before.

File: src/sssom/context.py
Comment (cthoyt) [reply src/sssom/context.py:103]: See also https://github.com/mapping-commons/sssom/pull/302 for a potential upstream fix. I don't think this should be a blocker for this PR, since prefix map loading isn't very principled in sssom-py already, I don't see any harm in adding some additional specific hacks. A future PR can switch over to using `curies.Converter` natively to handle prefix map loading and support synonyms.

Comment (matentzn) [reply src/sssom/context.py:103]: Yea, these entries should not exist in the sssom schema context. We want to move away from them entirely. 

I would prefer to wait for the ssssom schema release and then get rid of this dc superfluous code..

Comment (matentzn) [reply src/sssom/context.py:103]: This is for now necessary because the LinkML context generator automatically adds those prefixes.

File: src/sssom/context.py
Comment (matentzn) [inline src/sssom/context.py:130]: How does this line of code affect the rest of the code? It seems a bit disconnected?

File: tests/test_parsers.py
Comment (matentzn) [inline tests/test_parsers.py:67]: How does this line of code work?

File: src/sssom/context.py
Comment (cthoyt) [reply src/sssom/context.py:130]: it will throw an error if the prefix map is not bijective, I can update the comment to make this more clear

Comment (matentzn) [reply src/sssom/context.py:130]: Ah cool. No this makes sense now!

File: tests/test_parsers.py
Comment (matentzn) [reply tests/test_parsers.py:67]: Same as above.

Comment (cthoyt) [reply tests/test_parsers.py]: I added a more explicit helper function for checking prefix map validity in 439f561

File: src/sssom/context.py
Comment (cthoyt) [reply src/sssom/context.py]: I added a more explicit helper function for checking prefix map validity in 439f561

File: tests/test_parsers.py
Comment (cthoyt) [reply tests/test_parsers.py:128]: looking into it

File: .gitignore
Comment (cthoyt) [inline .gitignore:161]: @hrshdhgd if we're giving up on poetry locks, can we put that into a different PR? I don't want people saying later that I snuck this in :)

File: src/sssom/context.py
Comment (cthoyt) [reply src/sssom/context.py]: We actually fixed this the right way with the SSSOM Schema v0.14, I removed related code for checking/fixing DC/DCTERMS in e24fec2.

File: tests/test_parsers.py
Comment (cthoyt) [reply tests/test_parsers.py]: so yes, I have now added a detailed explanation of why this test is not needed. We should keep it skipped, or even better, remove it.

File: .gitignore
Comment (hrshdhgd) [reply .gitignore:161]: Fair enough. The tests will fail for now but I think I have a solution.

Comment (cthoyt) [reply .gitignore]: Yes, I mean we can solve the problem in a separate PR, merge it into master, then just update this PR on top of it. Thanks!

Comment (hrshdhgd) [reply .gitignore]: Done: https://github.com/mapping-commons/sssom-py/pull/399

**Comment by cthoyt** [comment]
@hrshdhgd it appears this PR is passing, even with the Poetry changes removed.

**Comment by hrshdhgd** [comment]
> @hrshdhgd it appears this PR is passing, even with the Poetry changes removed.

Yes , I had a quick solution that worked!

**Comment by hrshdhgd** [comment]
I merged #399 to main. Now if you merge the main branch down to this one, all should be well. I'll let you do it since you're actively working on this branch.

File: tests/test_parsers.py
Comment (matentzn) [inline tests/test_parsers.py:121]: @cthoyt I reinstated this test. Analysing the warnings, I noticed that the 800 less parsed entities (which is good, since they are not in the prefix map) are due to https://github.com/biopragmatics/bioregistry/issues/917. 

I think this is entirely independent of this PR, so I will approve this, this is just FYI.

**Review by matentzn — APPROVED** [review]
I have convinced myself that deleting that broken obographs test was the right thing to do! Thanks for your work on this, I am happy now; the one comment is purely an FYI, the behaviour is desireable.

**Comment by matentzn** [comment]
@hrshdhgd ready for your review and merge, I am happy with it!

**Review by hrshdhgd — APPROVED** [review]
Looks much cleaner than before! Thanks @cthoyt !

**Pull request merged**
